### PR TITLE
powertop: add package

### PR DESCRIPTION
**Pull request by lucize**
package for monitoring power consumption

depends on #8228 
would be also useful to have CONFIG_CPU_FREQ_STATS
Signed-off-by: Lucian Cristian <lucian.cristian@gmail.com>

Maintainer: me
Compile/Run tested: x86_64

### CODE REVIEW

**Comment by diizzyy** [comment]
LGTM
(Nitpick) You have inconsistent amount of new lines between segments is that intentional?

**Comment by lucize** [comment]
Yes, I saw it better like that, but I can remove them

File: utils/powertop/Makefile [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+PKG_BUILD_PARALLEL:=1
+PKG_INSTALL:=1
+
+include $(INCLUDE_DIR)/package.mk
+
+
+define Package/powertop
+  SECTION:=utils
+  CATEGORY:=Utilities
+  DEPENDS:=+libpci +libstdcpp +libncursesw +libnl-genl +zlib
```
Comment (neheb) [inline utils/powertop/Makefile]: does uclibc++ work here?

Comment (lucize) [reply utils/powertop/Makefile]:
```
/home/build/proxy/staging_dir/target-x86_64_musl/usr/include/uClibc++/algorithm: In instantiation of 'void std::sort(RandomAccessIterator, RandomAccessIterator, Compare) [with RandomAccessIterator = devuser**; Compare = bool (*)(devuser*, devuser*)]':
devlist.cpp:329:51:   required from here
/home/build/proxy/staging_dir/target-x86_64_musl/usr/include/uClibc++/algorithm:836:14: error: 'stable_sort' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
   stable_sort(first, last, comp);
   ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
```
and 


```
main.cpp: In function 'void one_measurement(int, int, char*)':
main.cpp:226:3: error: 'pthread_t' was not declared in this scope
   pthread_t thread = 0UL;
   ^~~~~~~~~
main.cpp:226:3: note: suggested alternative: 'pread'
   pthread_t thread = 0UL;
   ^~~~~~~~~
   pread
main.cpp:228:23: error: 'thread' was not declared in this scope
   if (pthread_create(&thread, NULL, measure_background_thread, &sample_interval))
                       ^~~~~~
main.cpp:228:23: note: suggested alternative: 'pread'
   if (pthread_create(&thread, NULL, measure_background_thread, &sample_interval))
                       ^~~~~~
                       pread
main.cpp:228:7: error: 'pthread_create' was not declared in this scope
   if (pthread_create(&thread, NULL, measure_background_thread, &sample_interval))
       ^~~~~~~~~~~~~~
main.cpp:228:7: note: suggested alternative: 'timer_create'
   if (pthread_create(&thread, NULL, measure_background_thread, &sample_interval))
       ^~~~~~~~~~~~~~
       timer_create
main.cpp:234:7: error: 'thread' was not declared in this scope
   if (thread)
       ^~~~~~
main.cpp:234:7: note: suggested alternative: 'pread'
   if (thread)
       ^~~~~~
       pread
main.cpp:237:4: error: 'pthread_join' was not declared in this scope
    pthread_join( thread, NULL);
    ^~~~~~~~~~~~
main.cpp:237:4: note: suggested alternative: 'thread_local'
    pthread_join( thread, NULL);
    ^~~~~~~~~~~~
    thread_local

```

**Comment by diizzyy** [comment]
It makes little sense to put in any effort on getting uclibc++ working as it most of the time offers subpar performance compared to libstdcpp and storage (<1Mbyte) isn't an issue on the platforms this is going to be used on.

File: utils/powertop/Makefile (hunk 1 above)
Comment (neheb) [reply utils/powertop/Makefile]: First issue is one that's fixed by my uclibc++ patch on OpenWrt's patchwork that everyone refuses to touch.

Second issue is a missing header.

**Comment by neheb** [comment]
Isn't powertop just a power measurement utility? Why would speed be important?

**Comment by diizzyy** [comment]
My point is that its probably not worth investing time on getting uclibc++ working just for the sake of it for storage size conerns given the platforms where powertop is usable. There are already quite a lot of package already depending on libstdcpp because of compatibility and/or peformance reasons and I would imagine that's the main reason we you'd run x86/x86_64 in the first place.

**Comment by neheb** [comment]
Ehhh no. Only two or three are using libstdcpp for that reason. The others because they use C++11 features not supported by uClibc++. I went through all the packages.

One issue is that uClibc++.mk has to be included. If not, dependency on libstdcpp shows up.

**Comment by diizzyy** [comment]
Let me rephrase, I don't see the need for a hard requirement of uclibc++ and spending time on it is up to the submitter. If it requries additional patching I'd rather see it in upstream.

File: utils/powertop/Makefile (hunk 1 above)
Comment (neheb) [reply utils/powertop/Makefile]: Just noticed as I'm building that this is a build requirement: https://github.com/openwrt/packages/pull/8228

Comment (lucize) [reply utils/powertop/Makefile]: it was stated in the first message

Comment (neheb) [reply utils/powertop/Makefile]: This builds fine locally. Looks like I need to submit a dumber version of the patch.

Comment (neheb) [reply utils/powertop/Makefile]: Ah false alarm. It fails later on as long long unsigned int is not supported with the >> operator. Although I thought I patched that locally... Will need to investigate.

For right now, uclibc++ is not possible.

**Comment by neheb** [comment]
Please number the patches, so that the order in which they should be applied is clear.

In other news, I got powertop to compile with uclibc++. Unfortunately it requires a patch to uclibc++, which nobody has been willing to touch.

**Comment by neheb** [comment]
I have two extra patches for you: https://github.com/neheb/packages/commit/de204dbcd0c77e393f59f3f1e9bc9692fe690b67

**Comment by diizzyy** [comment]
So drop uclibc++
Those patches are taken from Alpine Linux's repo as far as I can tell.
The commit message should reflect on that (origin of patches)

File: utils/powertop/Makefile [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#
+# Copyright (C) 2019 <lucian.cristian@gmail.com>
```
Comment (diizzyy) [inline utils/powertop/Makefile]: Real name

**Comment by neheb** [comment]
Note that there is a patch with which you can use to test uClibc++ on the mailing list: http://lists.infradead.org/pipermail/openwrt-devel/2019-April/016628.html

**Comment by lucize** [comment]
yes I saw that, but I'll update if it's merged. Thanks !

**Comment by neheb** [comment]
I mention it as it may not work correctly. Needs testing. In any case, I'll merge in a few minutes.
